### PR TITLE
InfoWindow using view annotations example

### PR DESCRIPTION
**Pull request by kiryldz**
### Summary of changes

Introduce new example activity almost replicating v9 maps [example](https://github.com/mapbox/mapbox-gl-native-android/blob/main/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/infowindow/InfoWindowActivity.java).

Also helper classes `Marker` and `MarkerManager` were introduced making example code almost look like in v9 or v6 maps. Those classes are **not** part of SDK, they are located in `:app` module. If users wish to use legacy `InfoWindow` functionality they are welcome to copy `Marker` / `MarkerManager` classes and adopt them as they wish. Functionality implemented now is:
 - use similar to v9 / v6 `InfoWindow` layout with either `title` or `snippet` or both
 - allow only one pop up be opened at a time - clicking anywhere closes it
 - adjusting X offset when opening an info window near left / right screen edge

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.

 | Maps v10 | Maps v9 |
 | - |  - |
 | <video src="https://user-images.githubusercontent.com/15800566/154271457-94c79031-b7cd-4ebd-a7c2-82f253e3eb99.mp4" width = 250/> | <video src="https://user-images.githubusercontent.com/15800566/154271363-1d685e5f-b1aa-4bab-bd06-e1aa89a6a19d.mp4" width = 250/> |

 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `./gradlew apiDump` to update generated api files, if there's public API changes, otherwise the `verify-kotlin-binary-compatibility` CI will fail.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog></changelog>`.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      R.drawable.blue_marker_view
+    )!!
+    mapView.getMapboxMap().apply {
+      setCamera(
+        cameraOptions {
+          center(Point.fromLngLat(-77.03655168667463, 38.897705003219784))
+          zoom(15.0)
+        }
+      )
+      loadStyleUri(Style.MAPBOX_STREETS) {
```
Comment (Chaoba) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:48]: Replace with `getStyle`

Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:48]: why?

Comment (Chaoba) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:48]: We would load `MAPBOX_STREETS` by default, no need to set it.

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import com.mapbox.maps.plugin.gestures.removeOnMapLongClickListener
+import com.mapbox.maps.testapp.R
+import com.mapbox.maps.testapp.utils.BitmapUtils
+import java.text.DecimalFormat
+
+/**
+ * Test activity showcasing using the legacy InfoWindow API above Washington D.C
+ * using new View Annotation API.
+ *
+ * @see [legacy example code](https://github.com/mapbox/mapbox-gl-native-android/blob/main/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/infowindow/InfoWindowActivity.java)
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:22]: do we want to link to legacy codebase? we didn't reference any of the legacy activities in our codebase.

Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:22]: Yeah, agree, will remove. Mentioning it in PR description should be enough (just to make it easier for us to navigate if needed)

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import com.mapbox.maps.plugin.gestures.removeOnMapClickListener
+import com.mapbox.maps.viewannotation.OnViewAnnotationPositionUpdatedListener
+import com.mapbox.maps.viewannotation.viewAnnotationOptions
+import java.util.concurrent.CopyOnWriteArrayList
+import kotlin.math.abs
+
+/**
+ * Manager class helping to control the [Marker]s.
+ */
+class MarkerManager(
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt:23]: would it make sense to move this class or similar abstraction to the view annotation(or as a separate plugin)?

Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt:23]: no, I've mentioned this point in PR description, we're leaving this as reference implementation for now 😉

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        // needed to display info window above the marker
+        offsetY(pointAnnotation.iconImageBitmap?.height!! + MARKER_PADDING_PX)
+      }
+    )
+    viewAnnotation.findViewById<TextView>(R.id.infowindow_title).text = title
+    viewAnnotation.findViewById<TextView>(R.id.infowindow_description).text = snippet
+    prepared = true
+  }
+
+  companion object {
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:64]:
```suggestion
  private companion object {
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+  internal fun prepareAnnotationMarker(pointAnnotationManager: PointAnnotationManager) {
+    val pointAnnotationOptions: PointAnnotationOptions = PointAnnotationOptions()
+      .withPoint(position)
+      .withIconImage(icon)
+      .withIconAnchor(IconAnchor.BOTTOM)
+    pointAnnotation = pointAnnotationManager.create(pointAnnotationOptions)
+  }
+
+  internal fun prepareViewAnnotation(mapView: MapView) {
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:45]: maybe better pass as argument `viewAnnotationManager`?

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    if (!marker.prepared) {
+      return
+    }
+    markerList.remove(marker)
+    mapView.viewAnnotationManager.removeViewAnnotation(marker.viewAnnotation)
+    pointAnnotationManager.delete(marker.pointAnnotation)
+  }
+
+  fun selectMarker(marker: Marker, deselectIfSelected: Boolean = false) {
+    if (mapView.viewAnnotationManager.getViewAnnotationOptionsByView(marker.viewAnnotation)?.selected == true) {
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt:71]: nit:

seems to be a good extension
`fun MapView.isMarkerSelected(marker) = viewAnnotationManager.getViewAnnotationOptionsByView(marker.viewAnnotation)?.selected == true`

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt (hunk 4 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:64]: I've used it in several places before and it's not convenient if some `public` or `internal` constant is about to be added - in that case we need to remove this modifier. Any other thoughts I'm missing why it may be useful comparing with current approach?

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt (hunk 5 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:45]: 👍

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  val snippet: String? = null,
+) {
+
+  init {
+    if (title == null && snippet == null) {
+      throw RuntimeException("Marker should have either title or snippet!")
+    }
+  }
+
+  internal var prepared = false
```
[the remark this answers is not in the document]
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:32]: 👍

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+      if (deselectIfSelected) {
+        deselectMarker(marker)
+      }
+      return
+    }
+    // Need to deselect any currently selected annotation first
+    deselectMarkers()
```
[the remark this answers is not in the document]
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt:78]: we have a check beforehand if selected marker == marker we want to select so if we hit  `deselectMarkers()` we will actually deselect other markers if they were selected. additionally I've copied this code from v9 maps to be in sync 😄

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt (hunk 6 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/MarkerManager.kt:71]: 👍

**Review by Chaoba — APPROVED** [review]
One nit, others LGTM

**Review by ank27 — APPROVED** [review]

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import com.mapbox.maps.extension.style.layers.properties.generated.IconAnchor
+import com.mapbox.maps.plugin.annotation.generated.PointAnnotation
+import com.mapbox.maps.plugin.annotation.generated.PointAnnotationManager
+import com.mapbox.maps.plugin.annotation.generated.PointAnnotationOptions
+import com.mapbox.maps.testapp.R
+import com.mapbox.maps.viewannotation.ViewAnnotationManager
+import com.mapbox.maps.viewannotation.viewAnnotationOptions
+
+/**
+ * Class representing a marker that contains marker icon and info window on top of it.
```
Comment (yunikkk) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:17]:
```suggestion
 * Class representing a marker that contains marker icon and ViewAnnotation on top of it.
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import com.mapbox.maps.plugin.annotation.generated.*
+import com.mapbox.maps.plugin.gestures.OnMapLongClickListener
+import com.mapbox.maps.plugin.gestures.addOnMapLongClickListener
+import com.mapbox.maps.plugin.gestures.removeOnMapLongClickListener
+import com.mapbox.maps.testapp.R
+import com.mapbox.maps.testapp.utils.BitmapUtils
+import java.text.DecimalFormat
+
+/**
+ * Test activity showcasing using the legacy InfoWindow API above Washington D.C
```
Comment (yunikkk) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:18]: Since there is not more InfoWindow API IMHO there's nothing to showcase. I'd reworded saying that `Instead of old InfoWindow API you might use View Annotation + Point Annotation APIs`

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ */
+data class Marker(
+  val position: Point,
+  val icon: Bitmap,
+  val title: String? = null,
+  val snippet: String? = null,
+) {
+
+  init {
+    if (title == null && snippet == null) {
```
Comment (alexander-kulikovskii) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:27]:
```suggestion
    if (title == null || snippet == null) {
```

**Review by alexander-kulikovskii — APPROVED** [review]

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt (hunk 10 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/InfoWindowActivity.kt:18]: agree

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt (hunk 9 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:17]: agree

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt (hunk 4 above)
Comment (alexander-kulikovskii) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:66]: yes, all of your constants inside `companion object` is private, but public `companion object` itself visible for all and contains nothing, that a little bit weird. and also wastes some space for end users.

but up to you

Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/infowindow/Marker.kt:66]: yeah, see your point, will change then, thanks for clarification!

**Review by yunikkk — APPROVED** [review]

**Review by pengdev — APPROVED** [review]
